chainstate/test-suite/src/tests/atomic_swap.rs

```diff
@@ -0,0 +1,617 @@
+                    OutputValue::Coin(Amount::from_atoms(100)),
+                    Destination::AnyoneCanSpend,
+                ))
+                .build()
+                .take_transaction();
+
+            let random_secret = HtlcSecret::new_from_rng(&mut rng);
+
+            let input_sign = produce_uniparty_signature_for_htlc_input(
+                &bob_sk,
+                SigHashType::try_from(SigHashType::ALL).unwrap(),
+                Destination::PublicKeyHash((&bob_pk).into()),
+                &tx,
+                &[Some(&tx_1.transaction().outputs()[0])],
+                0,
+                random_secret,
+            )
+            .unwrap();
+
+            let result = tf
+                .make_block_builder()
+                .add_transaction(
+                    SignedTransaction::new(tx, vec![InputWitness::Standard(input_sign)]).unwrap(),
+                )
+                .build_and_process();
+            assert_eq!(
+                result.unwrap_err(),
+                ChainstateError::ProcessBlockError(BlockError::StateUpdateFailed(
+                    ConnectTransactionError::SecretHashMismatch(UtxoOutPoint::new(
+                        tx_1_id.into(),
+                        0
+                    ))
+                ))
+            );
+        }
+
+        // Success case
+        let tx = TransactionBuilder::new()
+            .add_input(
+                TxInput::from_utxo(tx_1_id.into(), 0),
+                InputWitness::NoSignature(None),
+            )
+            .add_output(TxOutput::Transfer(
+                OutputValue::Coin(Amount::from_atoms(100)),
+                Destination::AnyoneCanSpend,
+            ))
+            .build()
+            .take_transaction();
+
+        let input_sign = produce_uniparty_signature_for_htlc_input(
+            &bob_sk,
+            SigHashType::try_from(SigHashType::ALL).unwrap(),
+            Destination::PublicKeyHash((&bob_pk).into()),
+            &tx,
+            &[Some(&tx_1.transaction().outputs()[0])],
+            0,
+            secret,
+        )
+        .unwrap();
+
+        tf.make_block_builder()
+            .add_transaction(
+                SignedTransaction::new(tx, vec![InputWitness::Standard(input_sign)]).unwrap(),
+            )
+            .build_and_process()
+            .unwrap();
+    });
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn refund_atomic_swap(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = test_utils::random::make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+
+        let chain_config = tf.chainstate.get_chain_config().clone();
+
+        let (alice_sk, alice_pk) = PrivateKey::new_from_rng(&mut rng, KeyKind::Secp256k1Schnorr);
+        let (bob_sk, bob_pk) = PrivateKey::new_from_rng(&mut rng, KeyKind::Secp256k1Schnorr);
+
+        let secret = HtlcSecret::new_from_rng(&mut rng);
+
```

Comment:
Can you make a function that acts as a fixture to produce the same settings and then spend it in both cases, refund and normal?